src/Streamly/Internal/FileSystem/Event/Windows.hs

```diff
@@ -0,0 +1,447 @@
+            bytesRet <- peekByteOff bret 0
+            readChangeEvents buffer root bytesRet
+ 
+type FileAction = DWORD
+
+fILE_ACTION_ADDED             :: FileAction
+fILE_ACTION_ADDED             =  1
+
+fILE_ACTION_REMOVED           :: FileAction
+fILE_ACTION_REMOVED           =  2
+
+fILE_ACTION_MODIFIED          :: FileAction
+fILE_ACTION_MODIFIED          =  3
+
+fILE_ACTION_RENAMED_OLD_NAME  :: FileAction
+fILE_ACTION_RENAMED_OLD_NAME  =  4
+
+fILE_ACTION_RENAMED_NEW_NAME  :: FileAction
+fILE_ACTION_RENAMED_NEW_NAME  =  5
+
+eventStreamAggr :: (HANDLE, FilePath, Config) -> SerialT IO Event
+eventStreamAggr (handle, rootPath, cfg) =  do
+    let recMode = getConfigRecMode cfg
+        flagMasks = getConfigFlag cfg
+    S.concatMap S.fromList $ S.repeatM 
+        $ readDirectoryChanges rootPath handle recMode flagMasks
+
+pathsToHandles ::  
+    NonEmpty FilePath -> Config -> SerialT IO (HANDLE, FilePath, Config)
+pathsToHandles paths cfg = do
+    let pathStream = S.fromList (NonEmpty.toList paths)
+        st2 = S.mapM getWatchHandle pathStream
+    S.map (\(h, f) -> (h, f, cfg)) st2   
+
+peekFNI :: Ptr FILE_NOTIFY_INFORMATION -> IO FILE_NOTIFY_INFORMATION
+peekFNI buf = do
+    neof <- peekByteOff buf 0
+    acti <- peekByteOff buf 4
+    fnle <- peekByteOff buf 8
+    fnam <- peekCWStringLen
+        -- start of array
+        (buf `plusPtr` 12,
+        -- fnle is the length in *bytes*, and a WCHAR is 2 bytes
+        fromEnum (fnle :: DWORD) `div` 2) 
+    return $ FILE_NOTIFY_INFORMATION neof acti fnam
+
+-------------------------------------------------------------------------------
+-- Utilities
+-------------------------------------------------------------------------------
+utf8ToString :: Array Word8 -> FilePath
+utf8ToString = runIdentity . S.toList . U.decodeUtf8 . A.toStream
+
+utf8ToStringList :: NonEmpty (Array Word8) -> NonEmpty FilePath
+utf8ToStringList = NonEmpty.map utf8ToString
+
+-- | Close a Directory handle.
+--
+-- /Internal/
+--
+closePathHandleStream :: SerialT IO (HANDLE, FilePath, Config) -> IO ()
+closePathHandleStream = S.mapM_ (\(h, _, _) -> closeHandle h)
+
+-- | Start monitoring a list of file system paths for file system events with
+-- the supplied configuration operation over the 'defaultConfig'. The
+-- paths could be files or directories. When the path is a directory, only the
+-- files and directories directly under the watched directory are monitored,
+-- contents of subdirectories are not monitored.  Monitoring starts from the
+-- current time onwards.
+--
+-- /Internal/
+--
+watchPathsWith :: 
+       (Config -> Config) 
+    -> NonEmpty (Array Word8) 
+    -> SerialT IO Event
+watchPathsWith f paths = do
+    let cfg = f $ setRecursiveMode False defaultConfig
+        sth = pathsToHandles (utf8ToStringList paths) cfg        
+    S.after (closePathHandleStream sth) 
+        $ S.concatMapWith parallel eventStreamAggr sth
```

Comment:
BUG: the stream `sth` is being evaluated multiple times. `after` and `concatMapWith` will independently evaluate the stream. We are closing different handles in `after` rather than the ones being used in `concatMapWith`.

Instead, we can use a stream of paths as input to `concatMapWith`, in eventStreamAggr generate a stream from the path it receives, it can use a `bracket` around it to open the path before and close it after the stream is done.

We have the same issue in watchTrees as well.